### PR TITLE
Only createOrFix client when necessary

### PR DESCRIPTION
**Pull request by angxu**
The current router implementation is causing high cpu/memory usage on muse root because in muse we always fanout to all shards and we have hundreds of hosts in each shard, as a result we will have to create a set of thousands hosts and iterate through it for each muse request. It seems we can avoid this problem by only `createOrFixClientsFor` hosts that we have selected to return.

### CODE REVIEW

File: common/thrift_router.h [hunk 1]
```diff
@@ -300,6 +288,7 @@ class ThriftRouter {
           ret = ReturnCode::NOT_FOUND;
           continue;
         }
+        createOrFixClientsFor(hosts_for_shard);
```
Comment (newpoo) [inline common/thrift_router.h]: Here we still do createOrFix for all hosts (hundreds of host in muse root) that host this shard, right?

Comment (angxu) [reply common/thrift_router.h]: good catch! I didn't realize `filterByRoleAndSortByPreference` also returns full list of hosts. But i think even just getting rid of the hash set above is a good optimization for us. I applied the current patch to production and it shows ~10% cpu usage and a few milliseconds latency reduction.

File: common/thrift_router.h [hunk 2]
```diff
@@ -382,25 +382,27 @@ class ThriftRouter {
             v.push_back(hi.first);
           }
         }
-        RankHostsByGroupPrefixLength(&v, rotation_counter, segment);
-        RankHostsByGroupPrefixLength(&v_s, rotation_counter, segment);
+        RankHostsByGroupPrefixLengthAndShrinkTo(&v, rotation_counter, segment, shrink_target);
+        RankHostsByGroupPrefixLengthAndShrinkTo(&v_s, rotation_counter, segment, shrink_target);
         v.insert(v.end(), v_s.begin(), v_s.end());
```
[the remark this answers is not in the document]
Comment (angxu) [reply common/thrift_router.h]: done.

File: common/thrift_router.h [hunk 3]
```diff
@@ -370,7 +359,8 @@ class ThriftRouter {
         const std::vector<std::pair<const Host*, Role>>& host_info,
```
[the remark this answers is not in the document]
Comment (angxu) [reply common/thrift_router.h]: done.

**Comment by newpoo** [comment]
awesome!
Feel free to check in

**Pull request merged**